src/Core/Content/Cms/DataResolver/CmsSlotsDataResolver.php

```diff
@@ -16,42 +17,52 @@
 use Shopware\Core\Framework\DataAbstractionLayer\Search\EntitySearchResult;
 use Shopware\Core\Framework\Struct\ArrayEntity;
 use Shopware\Core\System\SalesChannel\SalesChannelContext;
+use Symfony\Component\EventDispatcher\EventDispatcherInterface;
 
 class CmsSlotsDataResolver
 {
     /**
      * @var CmsElementResolverInterface[]
      */
-    private $resolvers;
+    private array                      $resolvers;
 
-    /**
-     * @var array
-     */
-    private $repositories;
+    private array                      $repositories;
 
-    /**
-     * @var DefinitionInstanceRegistry
-     */
-    private $definitionRegistry;
+    private DefinitionInstanceRegistry $definitionRegistry;
+
+    private EventDispatcherInterface   $eventDispatcher;
 
     /**
      * @param CmsElementResolverInterface[] $resolvers
      */
-    public function __construct(iterable $resolvers, array $repositories, DefinitionInstanceRegistry $definitionRegistry)
-    {
+    public function __construct(
+        iterable                   $resolvers,
+        array                      $repositories,
+        DefinitionInstanceRegistry $definitionRegistry,
+        EventDispatcherInterface   $eventDispatcher
+    ) {
         $this->definitionRegistry = $definitionRegistry;
+        $this->eventDispatcher    = $eventDispatcher;
 
         foreach ($repositories as $entityName => $repository) {
             $this->repositories[$entityName] = $repository;
         }
 
         foreach ($resolvers as $resolver) {
-            $this->resolvers[$resolver->getType()] = $resolver;
+            $this->addResolver($resolver);
         }
     }
 
+    public function addResolver(CmsElementResolverInterface $resolver): void
+    {
+        $this->resolvers[$resolver->getType()] = $resolver;
+    }
+
     public function resolve(CmsSlotCollection $slots, ResolverContext $resolverContext): CmsSlotCollection
     {
+        $event = new CmsSlotDataResolverEvent($this, $resolverContext->getSalesChannelContext());
```

Comment:
I guess you have added this event to call the `addResolver` method. Passing the whole class makes this event for me very risky and bad to move to another class.

What about another approach?

The constructor already expects an iterable. You could replace using a compiler pass the first argument with a service factory.

This service factory gets all current tagged services injected. You return an iterable object in the return (class needs to implement IteratorAggregate) in addition to your dynamic services.

This would work also in previous Shopware versions

Reply:
@tyurderi did it solve your problem?